features/filesystem/fat/FATFileSystem.cpp

```diff
@@ -123,6 +123,61 @@ void fat_filesystem_set_errno(FRESULT res)
     return;
 }
 
+// Helper class for deferring operations when variable falls out of scope
+template <typename T>
+class Deferred {
+public:
+    T _t;
+    Callback<void(T)> _ondefer;
+
+    Deferred(const Deferred&);
+    Deferred &operator=(const Deferred&);
+
+public:
+    Deferred(T t, Callback<void(T)> ondefer = NULL)
+        : _t(t), _ondefer(ondefer)
+    {
+    }
+
+    operator T()
+    {
+        return _t;
+    }
+
+    ~Deferred()
+    {
+        if (_ondefer) {
+            _ondefer(_t);
+        }
+    }
+};
+
+static void dodelete(const char *data)
+{
+    delete[] data;
+}
+
+// Adds prefix needed internally by fatfs, this can be avoided for the first fatfs
+// (id 0) otherwise a prefix of "id:/" is inserted in front of the string.
+static Deferred<const char*> fat_path_prefix(int id, const char *path)
```

Comment:
- The [fsid](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/fat/FATFileSystem.cpp#L264) will break.
- The [chanfs](https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/fat/FATFileSystem.cpp#L264) will fail to mount.
- [MBR](https://en.wikipedia.org/wiki/Master_boot_record) will fail with >4 partitions
- [GPT](https://en.wikipedia.org/wiki/GUID_Partition_Table) will fail with >128 partitions.
- You will need ~800KB of ram to mount all of the filesystems at once.
- With 160MB of storage you will be wasting 10% on metadata alone.

Oh and this code may break.